src/objective/multiclass_obj_gpu.cu

```diff
@@ -0,0 +1,245 @@
+                                            size_t ndata, size_t nclass) {
+  int i = threadIdx.x + blockIdx.x * blockDim.x;
+  if (i >= ndata)
+    return;
+  float *start = preds + i * nclass;
+  common::Softmax(start, start + nclass);
+}
+
+// Kernel for gradient and approximated hessian
+__global__ void multiclass_gradient_k(GradientPair *__restrict__ gpair,
+                                      float *__restrict__ labels,
+                                      float *__restrict__ preds,
+                                      float *__restrict__ weights,
+                                      size_t ndata, size_t nclass,
+                                      int *__restrict__ label_error) {
+  int i = threadIdx.x + blockIdx.x * blockDim.x;
+  if (i >= ndata)
+    return;
+
+  float *start = &preds[i * nclass];
+  common::Softmax(start, start + nclass);
+
+  auto label = labels[i];
+  if (label < 0 || label >= nclass) {
+    // If there is an incorrect label, the host code will know. No need to
+    // sync the threads.
+    *label_error = label;
+    label = 0;
+  }
+  bst_float wt = weights == nullptr ? 1.0f : weights[i];
+  for (int k = 0; k < nclass; ++k) {
+    bst_float p = *(start + k);
+    const float eps = 1e-16f;
+    const bst_float h = fmax(2.0f * p * (1.0f - p) * wt, eps);
+    if (label == k) {
+      gpair[i * nclass + k] = GradientPair((p - 1.0f) * wt, h);
+    } else {
+      gpair[i * nclass + k] = GradientPair(p * wt, h);
+    }
+  }
+}
+
+class GPUSoftmaxMultiClassObj : public ObjFunction {
+ protected:
+  void LazyResize(size_t n, size_t n_weights) {
+    if (labels_.Size() == n && weights_.Size() == n_weights) {
+      return;
+    }
+    copied_ = false;
+
+    labels_.Reshard(devices_);
+    weights_.Reshard(devices_);
+
+    if (labels_.Size() != n) {
+      labels_.Resize(n);
+    }
+    if (weights_.Size() != n_weights) {
+      weights_.Resize(n_weights);
+    }
+  }
+
+ public:
+  explicit GPUSoftmaxMultiClassObj(bool output_prob)
+      : output_prob_(output_prob), copied_{false} {}
+
+  void Configure(
+      const std::vector<std::pair<std::string, std::string>> &args) override {
+    param_.InitAllowUnknown(args);
+    CHECK(param_.n_gpus != 0) << "Must have at least one device";
+    devices_ = GPUSet::Range(param_.gpu_id, dh::NDevicesAll(param_.n_gpus));
+
+    max_data_.Reshard(devices_);
+  };
+
+  void GetGradient(HostDeviceVector<bst_float> *preds, const MetaInfo &info,
+                   int iter,
+                   HostDeviceVector<GradientPair> *out_gpair) override {
+    CHECK_NE(info.labels_.size(), 0U) << "label set cannot be empty";
+    CHECK(preds->Size() ==
+          (static_cast<size_t>(param_.num_class) * info.labels_.size()))
+        << "SoftmaxMultiClassObj: label size and pred size does not match";
+
+    const int nclass = param_.num_class;
+    size_t n = preds->Size();
+    auto ndata = static_cast<size_t>(n / nclass);  // number of data points
+
+    preds->Reshard(devices_);
```

Comment:
I see a potential error here e.g. if num_class is 7 and the first 3 predictions for a training instance end up on the first device and the last 4 predictions end up on the second device the result will be wrong.

This is not your fault we just did not think that far ahead with HostDeviceVector :)

cc @canonizer we could just add an optional integer parameter to the Reshard method indicating that data may not be separated below some granularity.

Reply:
Currently I can't contribute to the anything about multi-gpu. Since, well, I don't have multi-gpu. Hence, like the regression case, it's marked not implemented. 
But I will fix other stuffs soon. Thanks for the review. :)

Reply:
Hi. I tried to use lambda with `dh::LaunchN` inside the `Transform`, which generated the following error:
```error: The enclosing parent function ("Transform") for an extended  __device__ lambda cannot have private or protected access within its class```

There is a problem with closure. Is it really necessary to use this helper function? Making a work around might make things quirky.

In the last commit (except for further commit when fixing clang-tidy is needed), I tried to define BLK_SIZE to represent the default block size for CUDA in xgboost (which is 256). With that, although I didn't use dh helper function, information duplication should not be increased because of a new file. :-)  If you don't like it or want it to be in a separated PR, I can revert that commit.

Reply:
Let's keep the kblock changes for a different PR - though this change does seem reasonable to me. Note also that the kAlign parameter is not related to block size but is a number of bytes so probably shouldn't be assigned as kBlock. Using dh::LaunchN was more of a suggestion, what you have now is fine.

Reply:
The last commit is removed.